### PR TITLE
[tx] Paged Attention in SkyRL-tx

### PR DESCRIPTION
**Pull request by aditchawdhary**
Adding paged attention implementation in paged attention in skyrl-tx

### CODE REVIEW

File: skyrl-tx/tx/utils/paged_attention.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    if attention_mask is not None:
+        # Expand mask to match attention scores shape
+        mask_expanded = attention_mask[:, None, :, :]  # [B, 1, T, T]
+        scores = jnp.where(mask_expanded, scores, -1e9)
+    
+    # Softmax and weighted sum
+    attn_weights = jax.nn.softmax(scores, axis=-1)
+    output = jnp.einsum('bhtT,bThd->bthd', attn_weights, v)
+    
+    return output
```
[the remark this answers is not in the document]
Comment (aditchawdhary) [reply skyrl-tx/tx/utils/paged_attention.py]: thanks gemini bot! ❤️

File: skyrl-tx/tx/utils/paged_attention.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        )
+        
+        return (k_pages, v_pages), None
+    
+    page_offsets_for_seqs = cache.page_offsets[seq_indices]
+    (k_pages_updated, v_pages_updated), _ = jax.lax.scan(
+        write_sequence,
+        (cache.k_pages, cache.v_pages),
+        (seq_indices, k_new, v_new, page_offsets_for_seqs)
+    )
```
[the remark this answers is not in the document]
Comment (aditchawdhary) [reply skyrl-tx/tx/utils/paged_attention.py]: thanks gemini ❤️ !

File: skyrl-tx/tx/models/qwen3.py [hunk 3]
```diff
@@ -95,23 +96,36 @@ def __call__(
...
+            # Paged attention path
+            seq_indices = jnp.arange(B)
+            attn_output, updated_cache = paged_attention_with_update(
+                q=q,
+                k_new=k,
+                v_new=v,
+                cache=kv_cache,
+                seq_indices=seq_indices,
+                scale=1.0 / self.head_dim**0.5,
+            )
```
[the remark this answers is not in the document]
Comment (aditchawdhary) [reply skyrl-tx/tx/models/qwen3.py]: thanks gemini bot! ❤️

**Comment by aditchawdhary** [comment]
hey @pcmoritz, can you take a look and review the changes

**Comment by pcmoritz** [comment]
Thanks a lot for getting this started, I already took a first look -- I'm currently busy with getting the FSDP PR merged and will have a much closer look once that's done. Here are some high level comments of things I'm thinking about:

- In my mind, the primary motivation to implement paged attention is to implement continuous batching (see e.g. https://www.anyscale.com/blog/continuous-batching-llm-inference), and the main motivation for that is if you have requests of very different lengths, you can return the requests that are already finished and return them, so their slots can be taken up by new requests. This can both lead to computational savings (a lot of the request length distribution has high variance and currently in that setting we would do a decent amount of padding) and also faster responses. There is some secondary motivation of making sure other features like prefix caching etc. are building on the right foundations (and there might be more). So we should think a little more about what the best way to implement continuous batching will be. In the multi-tenancy setting there could be a lot of interesting questions on how it interacts with training (e.g. sampling prefill and forward / backward passes have very similar computational characteristics and could be scheduled together), but we can ignore all of that to start off with and just think about the single tenant sampling -> training -> sampling -> training -> ... setting and see how that can be made more efficient in the context of large variance in request length.
- Going forward, I think we should have only one format of KV cache for the tx native engine that works well. Most likely a paged one and we should study which data structures make the most sense (in my experience, deciding on the data representation is the most important and the code follows). In particular, we should study both sglang and vllm.
- It is also good to study which GPU / TPU kernels are available (e.g. @atemaguer pointed out https://github.com/sgl-project/sglang-jax/blob/main/python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py). Most likely we will need some sort of abstraction for the attention layer going forward, since there are many different kernels and techniques. Probably not the highest priority right now.

Anyways, these don't need to happen all at once and it is good to start with something small. If you want to do some more stuff in the mean time, I would encourage you to think a little more about the continuous batching problem which is a very real performance optimization, see also https://github.com/NovaSky-AI/SkyRL/pull/769 which remediates the need a little bit (but doesn't solve the high variance in length distribution problem). We also need to implement early stopping etc. And your approach of copying together the chunks in the kv cache and then using a vanilla attention implementation makes a lot of sense as the first step for that (have you measured the performance impact of that btw?).

Let's discuss more in this thread, I would love to hear your thoughts about these things :)

**Comment by aditchawdhary** [comment]
I looked in to the repository, [SGLang-JAX](https://github.com/sgl-project/sglang-jax), and they are using memory cache and kernels to implement KV caching, their kernels are specific to TPUs. I have created a [document](https://docs.google.com/document/d/1wQeRjhzug-iLMY_VLWmTqLOCj8bCJXy93V8ziR8Fvxs/edit?usp=sharing) for this I tried to integrate those with tinker and the inference results were 20% slower.
```
Current Benchmark:
python benchmarks/benchmark_engine.py   --base-model Qwen/Qwen3-0.6B   --benchmark sample   --num-steps 5   --num-requests 4   --seq-len 128   --sample-max-tokens 64
=== Sampling Benchmark ===
Warming up (2 steps)...
INFO:     JIT compiling for sample seq_len=128 in progress...
INFO:     JIT compilation for sample seq_len=128 took 34.50s
Running benchmark (5 steps)...

Results:
  steps:                5
  elapsed:              1.954 s
  tokens generated/sec: 655
  sec/step:              0.39
```
when I did the integration I was getting about 480 token generated per/s.

When I was going through skyrl-tx engine I realized that we can just use vllm/sglang directly for inference and that should give us better performance than trying to implement KV caching.

And there is a placeholder for this in Tinker, so I implemented the external inference, there results are better with 920 tokens generated/sec.
```
(main) root@C.29297665:/workspace$ CUDA_VISIBLE_DEVICES=0 vllm serve Qwen/Qwen3-0.6B --port 8000

(main) root@C.29297665:/workspace/skyrl-tx$ CUDA_VISIBLE_DEVICES=1 python benchmarks/benchmark_engine.py   --base-model Qwen/Qwen3-0.6B   --benchmark sample   --external-inference-url http://127.0.0.1:8000   --num-steps 5   --num-requests 4   --seq-len 128   --sample-max-tokens 64

=== Sampling Benchmark ===
Warming up (2 steps)...
Running benchmark (5 steps)...

Results:
  steps:                5
  elapsed:              1.392 s
  tokens generated/sec: 920
  sec/step:              0.28
```
